hw: Make device emulation files configurable

yangzhon opened the pull request:
@sameo @rbradford @mcastelino

This PR make all files in hw/ directory configurable and make new x86_64_virt-softmmu.mak for virt-i386 platform. With this new platform mak file, we can make specific binary for virt platform.

How to build binary with this new mak file
./configure --target-list=x86_64_virt-softmmu
or ref below sameo's config script. 

The FULL build will omit x86_64_virt-softmmu build and virt platform will be still compiled into x86_64-softmmu binary.

Now, with this PR, the full build or x86_64_virt-softmmu build all are successful.

# Default configuration for x86_64_virt-softmmu
CONFIG_VHOST_USER_SCSI=$(call land,$(CONFIG_VHOST_USER),$(CONFIG_LINUX))
CONFIG_VHOST_USER_BLK=$(call land,$(CONFIG_VHOST_USER),$(CONFIG_LINUX))

CONFIG_PCI=y
CONFIG_SCSI=y
CONFIG_NVME_PCI=y

CONFIG_VIRTIO=y
CONFIG_VIRTIO_9P=y
CONFIG_VIRTIO_BALLOON=y
CONFIG_VIRTIO_BLK=y
CONFIG_VIRTIO_CRYPTO=y
CONFIG_VIRTIO_NET=y
CONFIG_VIRTIO_RNG=y
CONFIG_VIRTIO_SCSI=y
CONFIG_VIRTIO_SERIAL=y
CONFIG_VIRTIO_PCI=y

CONFIG_ACPI=y
CONFIG_ACPI_X86=y
CONFIG_APIC=y
CONFIG_IOAPIC=y
CONFIG_VTD=y
CONFIG_AMD_IOMMU=y
CONFIG_SYS_BUS_DEBUG=y
CONFIG_PCI_LITE=y
CONFIG_ACPI_HW_REDUCED=y
CONFIG_X86_VIRT=y
CONFIG_MEM_HOTPLUG=y
CONFIG_ACPI_MEMORY_HOTPLUG=y
CONFIG_ACPI_CPU_HOTPLUG=y
CONFIG_ACPI_PCI_HOTPLUG=y

CONFIG_PFLASH_CFI01=y
CONFIG_HARD_DISK=y

**#The config for compile virt**
#!/bin/bash
set -e

if [[ "$EXTRA_CFLAGS" == "" ]]; then
    EXTRA_CFLAGS=" -O3 -fno-semantic-interposition -falign-functions=32 -D_FORTIFY_SOURCE=2 -fPIE"
fi

../configure \
 --disable-fdt \
 --disable-libiscsi \
 --disable-libnfs \
 --disable-libssh2 \
 --disable-linux-aio \
 --disable-lzo \
 --disable-modules \
 --disable-netmap \
 --disable-qom-cast-debug \
 --disable-seccomp \
 --disable-snappy \
 --disable-tcmalloc \
 --disable-tools \
 --disable-tpm \
 --disable-virtfs \
 --disable-tcg \
 --disable-slirp \
 --disable-capstone \
 --disable-usb-redir \
 --disable-xen \
 --disable-wdt \
 --enable-attr \
 --enable-cap-ng \
 --enable-kvm \
 --enable-rbd \
 --enable-vhost-crypto \
 --enable-vhost-net \
 --enable-vhost-scsi \
 --enable-vhost-user \
 --enable-vhost-vsock \
 --enable-virtfs \
 --target-list=x86_64_virt-softmmu \
 --extra-cflags="$EXTRA_CFLAGS" \
 --extra-ldflags=" -pie -z noexecstack -z relro -z now" \
 --libdir=/usr/lib64/nemu \
 --libexecdir=/usr/libexec/nemu \
 --datadir=/usr/share/nemu



As for hotplug functions configurable mentioned by @mcastelino, i will implement this in next PR.

Please help review this PR, any comments are welcome!

yangzhon committed:
hw: block: Make cdrom build optional

The cdrom.o object only depends on SCSI and IDE, so it should only be
built when those ones are enabled.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: i2c: change CONFIG definition for ich9

The SMBUS ICH9 (q35) controller only needs to be built when the new
ACPI_X86_ICH configuration knob is turned on. Without this we include
this emulated controller as part of the default x86_64 build.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: pci-host: make pam build configurable

pam is only used by pc/q35, so make it build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: usb: make core code configurable

We don't necesarily need core USB code, so we make it optional through
CONFIG_USB_CORE.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: softmmu: make new x86_64_virt-softmmu.mak

Add a new virt specific configuration file, with a minimized set of
features and devices enabled.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: Remove q35/pc code from the virt platform configuration

Remove the CONFIGs in the x86_64_virt-softmmu.mak and
compile out pc/q35 from it.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
configure: Remove virt from the full build

virt uses a very minimal set of features that breaks the full x86 build.
We remove it from the full build but keep the pc and q35 machine types
from the virt one.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: display: make VGA build configurable

Add a few configuration knobs in order to make the framebuffer, VGA and
display code configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon commented:
@sameo ,

The build is almost be successfully done now and there are two issues with FULL build and virt target build
(1) FULL build  ---> this issue is caused by moving fw_build_smbios() to hw/smbios/smbios.c file
/home/chen/94/nemu/hw/smbios/smbios.c:1140:5: error: unknown type name ‘X86CPU’; did you mean ‘CPU’?
     X86CPU *cpu = X86_CPU(ms->possible_cpus->cpus[0].cpu);
     ^~~~~~
     CPU
/home/chen/94/nemu/hw/smbios/smbios.c:1140:19: error: implicit declaration of function ‘X86_CPU’; did you mean ‘BP_CPU’? [-Werror=implicit-function-declaration]
     X86CPU *cpu = X86_CPU(ms->possible_cpus->cpus[0].cpu);

(2) virt build
../qapi/qapi-commands-run-state.o: In function `qmp_marshal_watchdog_set_action':
/home/chen/nemu/build/qapi/qapi-commands-run-state.c:104: undefined reference to `qmp_watchdog_set_action'

Commit:
usb: Add HMP stub routine

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

sameo commented:
@yangzhon With the following configure script:

```shell
#!/bin/bash
set -e

if [[ "$EXTRA_CFLAGS" == "" ]]; then
    EXTRA_CFLAGS=" -O3 -fno-semantic-interposition -falign-functions=32 -D_FORTIFY_SOURCE=2 -fPIE"
fi

../configure \
 --disable-fdt \
 --disable-libiscsi \
 --disable-libnfs \
 --disable-libssh2 \
 --disable-linux-aio \
 --disable-lzo \
 --disable-modules \
 --disable-netmap \
 --disable-qom-cast-debug \
 --disable-seccomp \
 --disable-snappy \
 --disable-tcmalloc \
 --disable-tools \
 --disable-tpm \
 --disable-virtfs \
 --disable-tcg \
 --disable-slirp \
 --disable-capstone \
 --disable-usb-redir \
 --disable-xen \
 --disable-wdt \
 --enable-attr \
 --enable-cap-ng \
 --enable-kvm \
 --enable-rbd \
 --enable-vhost-crypto \
 --enable-vhost-net \
 --enable-vhost-scsi \
 --enable-vhost-user \
 --enable-vhost-vsock \
 --enable-virtfs \
 --target-list=x86_64_virt-softmmu \
 --extra-cflags="$EXTRA_CFLAGS" \
 --extra-ldflags=" -pie -z noexecstack -z relro -z now" \
 --libdir=/usr/lib64/nemu \
 --libexecdir=/usr/libexec/nemu \
 --datadir=/usr/share/nemu

```

The build passes.

Commit:
hw: i386: virt: Use virt specific MADT CPU entry routine

acpi-build.c being built out, we need to have our own.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

sameo commented:
With a few ugly fixes, I can build and boot a CL image on top of virt. The new device model still needs some work, but looks saner than the vanilla one:

```
(qemu) info qdm
Controller/Bridge/Hub devices:
name "pci-lite-host", bus System, no-user
name "vfio-pci-igd-lpc-bridge", bus PCI, desc "VFIO dummy ISA/LPC bridge for IGD assignment"

Storage devices:
name "cfi.pflash01", bus System, no-user
name "nvme", bus PCI, desc "Non-Volatile Memory Express"
name "scsi-block", bus SCSI, desc "SCSI block device passthrough"
name "scsi-cd", bus SCSI, desc "virtual SCSI CD-ROM"
name "scsi-disk", bus SCSI, desc "virtual SCSI disk or CD-ROM (legacy)"
name "scsi-generic", bus SCSI, desc "pass through generic scsi device (/dev/sg*)"
name "scsi-hd", bus SCSI, desc "virtual SCSI disk"
name "vhost-scsi", bus virtio-bus
name "vhost-scsi-pci", bus PCI
name "vhost-user-blk", bus virtio-bus
name "vhost-user-blk-pci", bus PCI
name "vhost-user-scsi", bus virtio-bus
name "vhost-user-scsi-pci", bus PCI
name "virtio-9p-device", bus virtio-bus
name "virtio-9p-pci", bus PCI, alias "virtio-9p"
name "virtio-blk-device", bus virtio-bus
name "virtio-blk-pci", bus PCI, alias "virtio-blk"
name "virtio-scsi-device", bus virtio-bus
name "virtio-scsi-pci", bus PCI, alias "virtio-scsi"

Network devices:
name "virtio-net-device", bus virtio-bus
name "virtio-net-pci", bus PCI, alias "virtio-net"

Input devices:
name "virtconsole", bus virtio-serial-bus
name "virtio-input-host-device", bus virtio-bus
name "virtio-input-host-pci", bus PCI, alias "virtio-input-host"
name "virtio-keyboard-device", bus virtio-bus
name "virtio-keyboard-pci", bus PCI, alias "virtio-keyboard"
name "virtio-mouse-device", bus virtio-bus
name "virtio-mouse-pci", bus PCI, alias "virtio-mouse"
name "virtio-serial-device", bus virtio-bus
name "virtio-serial-pci", bus PCI, alias "virtio-serial"
name "virtio-tablet-device", bus virtio-bus
name "virtio-tablet-pci", bus PCI, alias "virtio-tablet"
name "virtserialport", bus virtio-serial-bus

Display devices:
name "virtio-gpu-device", bus virtio-bus
name "virtio-gpu-pci", bus PCI, alias "virtio-gpu"

Misc devices:
name "sysbus-debugcon", bus System
name "vfio-pci", bus PCI, desc "VFIO-based PCI device assignment"
name "vhost-vsock-device", bus virtio-bus
name "vhost-vsock-pci", bus PCI
name "virtio-balloon-device", bus virtio-bus
name "virtio-balloon-pci", bus PCI, alias "virtio-balloon"
name "virtio-crypto-device", bus virtio-bus
name "virtio-crypto-pci", bus PCI
name "virtio-rng-device", bus virtio-bus
name "virtio-rng-pci", bus PCI, alias "virtio-rng"

Uncategorized devices:
name "amd-iommu", bus System
name "AMDVI-PCI", bus PCI
name "apic", no-user
name "fw_cfg_io", bus System, no-user
name "fw_cfg_mem", bus System, no-user
name "intel-iommu", bus System
name "ioapic", bus System, no-user
name "kvm-apic", no-user
name "kvm-ioapic", bus System, no-user
name "kvmclock", bus System, no-user
name "kvmvapic", bus System, no-user
name "loader", desc "Generic Loader"
name "migration", no-user
name "mmio_interface", no-user
name "nvdimm", desc "DIMM memory module"
name "or-irq", no-user
name "pc-dimm", desc "DIMM memory module"
name "pci-lite-device", bus PCI, desc "Host bridge", no-user
name "qemu,register", no-user
name "split-irq", no-user
name "virt-acpi", bus System, desc "ACPI", no-user
```

cc @rbradford @mcastelino @sboeuf

sameo commented on hw/i386/virt/stub.c:
```diff
@@ -25,3 +26,16 @@ int pic_get_output(DeviceState *d)
+int vfio_display_probe(VFIOPCIDevice *vdev, Error **errp)
+{
+    return 0;
+}
```
Those stubs should be defined e.g. under `hw/vfio/stub.c` and only built when `CONFIG_VFIO_VGA` is not defined. They are not virt specific, and thus will be needed by anyone disabling `CONFIG_VFIO_VGA`.

yangzhon replied:
@sameo 

Yes, thanks, i will do ti.

sameo commented:
@yangzhon A few comments on this PR:

1. We need to move all the stubs into their own subsystem. See e.g. https://github.com/intel/nemu/pull/80/commits/1be5d66369f13d64c52b0b6bbe48ec765a95b853#r216498770.

2. We need to keep only non platform specific ACPI routines into `hw/i386/acpi-build.c` and move machine type specific ones into their respective machine types files. We should always build acpi-build.c.

3. When looking at the resulting device model, I see a few devices that could be still removed:

```
name "vfio-pci-igd-lpc-bridge", bus PCI, desc "VFIO dummy ISA/LPC bridge for IGD assignment"
name "virtio-input-host-device", bus virtio-bus
name "virtio-input-host-pci", bus PCI, alias "virtio-input-host"
name "virtio-keyboard-device", bus virtio-bus
name "virtio-keyboard-pci", bus PCI, alias "virtio-keyboard"
name "virtio-mouse-device", bus virtio-bus
name "virtio-mouse-pci", bus PCI, alias "virtio-mouse"
name "virtio-tablet-device", bus virtio-bus
name "virtio-tablet-pci", bus PCI, alias "virtio-tablet"
name "virtserialport", bus virtio-serial-bus
```

So we should also add some patches to configure them out.

4. Some devices presence should be discussed:

```
name "cfi.pflash01", bus System, no-user
name "scsi-block", bus SCSI, desc "SCSI block device passthrough"
name "scsi-cd", bus SCSI, desc "virtual SCSI CD-ROM"
name "scsi-disk", bus SCSI, desc "virtual SCSI disk or CD-ROM (legacy)"
name "scsi-generic", bus SCSI, desc "pass through generic scsi device (/dev/sg*)"
name "scsi-hd", bus SCSI, desc "virtual SCSI disk"
name "virtio-gpu-device", bus virtio-bus
name "virtio-gpu-pci", bus PCI, alias "virtio-gpu"
```

mcastelino commented on default-configs/x86_64_virt-softmmu.mak:
```diff
@@ -0,0 +1,43 @@
+CONFIG_PCI_LITE=y
+CONFIG_ACPI_HW_REDUCED=y
+CONFIG_X86_VIRT=y
+CONFIG_MEM_HOTPLUG=y
```
Does hotplug have to be enabled by default?

yangzhon replied:
@mcastelino,

If we also need compile out CPU and ACPI PCI hotplug? If like this, we will not support any hotplug. thanks!

rbradford replied:
@mcastelino Let's worry about a virt-minimal default-config file later.

mcastelino commented on default-configs/x86_64_virt-softmmu.mak:
```diff
@@ -0,0 +1,43 @@
+CONFIG_ACPI_X86=y
+CONFIG_APIC=y
+CONFIG_IOAPIC=y
+CONFIG_NVDIMM=y
```
Same with NVDIMM. Do we need this to be enabled by default.

yangzhon commented:
@sameo 

name "kvmvapic", bus System, no-user

We do not need kvmvapic, which is only used in older amd chips and older windows xp system.

yangzhon commented:
@sameo @rbradford @mcastelino




We need to define minimal device models for virt platform, and we can do porting per this target.

By the way, the display in VFIO and some legacy devices in virtio will need split files to get those device models compiled out.

The total device models in virt platform is around 67, the device model in Microsoft is around 20, the gap is 47.

yangzhon commented:
@sameo 

As for 2) 
We need to keep only non platform specific ACPI routines into hw/i386/acpi-build.c and move machine type specific ones into their respective machine types files. We should always build acpi-build.c.

In fact, only build_madt() in hw/i386/acpi-build.c will be called by virt platform and other code are related with PC and Q35. whether need we split this to three files ? acpi-build-core.c, acpi-build-piix.c and acpi-build-q35.c ?  which can make the code more cleaner if we do virt build. thanks!

sameo commented:
@yangzhon 
> acpi-build-core.c, acpi-build-piix.c and acpi-build-q35.c ? which can make the code more cleaner if we do virt build. thanks!

I'm fine with that option as well.

sameo commented:
@yangzhon 
> The total device models in virt platform is around 67

We have most of the virtio devices duplicated between `virtio-foo` and `virtio-foo-pci`. Not sure how we should count here.

rbradford commented:
@sameo count them once as the non -pci suffix version is just an alias.

sameo commented:
@yangzhon 
> We do not need kvmvapic, which is only used in older amd chips and older windows xp system.

When booting with CONFIG_KVM_VAPIC turned off:

```
qemu-system-x86_64_virt: Unknown device 'kvmvapic' for default sysbus
```

I'll investigate further.

sameo commented:
@yangzhon We're down to 46 when removing the PCI aliases.

yangzhon commented:
@sameo @rbradford @mcastelino

In the upstreamed qemu:
one by one to count: total 245 device model.
count by rob's method:  228 device model.

yangzhon commented:
#define TYPE_VIRTIO_CONSOLE_SERIAL_PORT "virtserialport"
static const TypeInfo virtconsole_info = {
        .name          = "virtconsole",
        .parent        = TYPE_VIRTIO_CONSOLE_SERIAL_PORT,
        .class_init    = virtconsole_class_init,
};

So we need keep virtserialport device model
name "virtserialport", bus virtio-serial-bus

The last virtio input device list
Input devices:
name "virtconsole", bus virtio-serial-bus
name "virtio-serial-device", bus virtio-bus
name "virtio-serial-pci", bus PCI, alias "virtio-serial"
name "virtserialport", bus virtio-serial-bus

yangzhon committed:
vfio: display: make vfio display configurable

If we disable vfio display, we need to add stub functions
for some display function used in vfio/pci.c.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon commented:
@sameo,

The total device model number related IRQ is 7,
name "apic", no-user
name "ioapic", bus System, no-user
name "kvm-apic", no-user
name "kvm-ioapic", bus System, no-user
name "kvmvapic", bus System, no-user
name "or-irq", no-user
name "split-irq", no-user

below device model have been compiled out by me and samuel
name "kvmvapic", bus System, no-user
name "or-irq", no-user
name "split-irq", no-user

yangzhon commented:
@sameo 

more patches uploaded.

Now the final info qdm as below
Controller/Bridge/Hub devices:
name "pci-lite-host", bus System, no-user

Storage devices:
name "cfi.pflash01", bus System, no-user
name "nvme", bus PCI, desc "Non-Volatile Memory Express"
name "scsi-block", bus SCSI, desc "SCSI block device passthrough"
name "scsi-cd", bus SCSI, desc "virtual SCSI CD-ROM"
name "scsi-disk", bus SCSI, desc "virtual SCSI disk or CD-ROM (legacy)"
name "scsi-generic", bus SCSI, desc "pass through generic scsi device (/dev/sg*)"
name "scsi-hd", bus SCSI, desc "virtual SCSI disk"
name "vhost-scsi", bus virtio-bus
name "vhost-scsi-pci", bus PCI
name "vhost-user-blk", bus virtio-bus
name "vhost-user-blk-pci", bus PCI
name "vhost-user-scsi", bus virtio-bus
name "vhost-user-scsi-pci", bus PCI
name "virtio-9p-device", bus virtio-bus
name "virtio-9p-pci", bus PCI, alias "virtio-9p"
name "virtio-blk-device", bus virtio-bus
name "virtio-blk-pci", bus PCI, alias "virtio-blk"
name "virtio-scsi-device", bus virtio-bus
name "virtio-scsi-pci", bus PCI, alias "virtio-scsi"

Network devices:
name "virtio-net-device", bus virtio-bus
name "virtio-net-pci", bus PCI, alias "virtio-net"

Input devices:
name "virtconsole", bus virtio-serial-bus
name "virtio-serial-device", bus virtio-bus
name "virtio-serial-pci", bus PCI, alias "virtio-serial"
name "virtserialport", bus virtio-serial-bus

Misc devices:
name "sysbus-debugcon", bus System
name "vfio-pci", bus PCI, desc "VFIO-based PCI device assignment"
name "vhost-vsock-device", bus virtio-bus
name "vhost-vsock-pci", bus PCI
name "virtio-balloon-device", bus virtio-bus
name "virtio-balloon-pci", bus PCI, alias "virtio-balloon"
name "virtio-crypto-device", bus virtio-bus
name "virtio-crypto-pci", bus PCI
name "virtio-rng-device", bus virtio-bus
name "virtio-rng-pci", bus PCI, alias "virtio-rng"

Uncategorized devices:
name "amd-iommu", bus System
name "AMDVI-PCI", bus PCI
name "apic", no-user
name "fw_cfg_io", bus System, no-user
name "fw_cfg_mem", bus System, no-user
name "intel-iommu", bus System
name "ioapic", bus System, no-user
name "kvm-apic", no-user
name "kvm-ioapic", bus System, no-user
name "kvmclock", bus System, no-user
name "loader", desc "Generic Loader"
name "migration", no-user
name "mmio_interface", no-user
name "pc-dimm", desc "DIMM memory module"
name "pci-lite-device", bus PCI, desc "Host bridge", no-user
name "virt-acpi", bus System, desc "ACPI", no-user

The total device model number in my side is 40.

yangzhon committed:
hw: minimize x86_64_virt-softmmu.mak

This is the minimal config file for the virt machine type.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: intc: compile out i8259

When i8259 get compiled out, we need to add a few
stubs for two specific functions used in apic.c.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: watchdog: compile out watchdog

We add watchdog stub functions together with a set of watchdog disabling
options for adding CONFIG_WDT to config-host.mak

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw : timer: compile rtc timer out

If we disable rtc, we need to add stub function
for the QMP rtc support.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: i386: compile kvmvapic out

Add stub functions for allowing kvmvapic to be compiled out.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon commented:
@sameo,

The updated PR include
(1) Rebase to latest code
(2) Remove hw/i386/virt/stub.c file and move stub function to each directory in hw/

Commit:
hw: vfio: Move quirks into their own directory

For configurability purposes, we're going to split pci-quirks into
smaller pieces. This is more cleanly done under its own sub-directory.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: vfio: quirks: Extract IGD code into its own file

We want to be able to compile out all or part of the VFIO PCI quirks, so
we need to put each and everyone of them into their own file.
We start with the IGD code.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

sameo commented:
@yangzhon I just pushed 3 commits to make the IGD in particular and the VFIO PCI quirks in general configurable as well. This is a WIP as it only extracts IGD out for now, but qdm shows the device is gone.
I need to add the new VFIO_PCI_QUIRKS default options to hw-common.mak.

yangzhon committed:
hw: i386: move build_madt to acpi-build-common.c

Move build_madt() to acpi-build-common.c file, in order to be able to
compile acpi-build only for non HW reduced machine types (pc and q35).

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: i386: split pc.c file

Split the i386/pc.c file between pc/q35 specific and x86 common code.
This is the only way to not build the entire pc core code for the non
emulated platforms like virt.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: i386: compile multiboot out

stub functions needed once multiboot gets compiled out.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
hw: vfio: quirks: Split Realtek quirks out

And provide stubs when VFIO_QUIRKS_REALTEK is not set.
This way we can potentially reduce the amount of
VFIO PCI quirks that are built in.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

yangzhon committed:
hw: i386: compile smbios related code out

smbios code can be compiled out when we disable CONFIG_SMBIOS.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: usb: compile out whole USB

Add related stub functions to compile out whole usb.
Setting the CONFIG_USB_CORE=y into usb.mak so that
the full build still passes.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
hw: vfio: quirks: Allow for IGD quirks to be compiled out

We create a stub file for IGD, and it's only built when
VFIO_PCI_QUIRKS_IGD is not set.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: vfio: quirks: Split ATI quirks out

And provide stubs when VFIO_QUIRKS_ATI is not set.
This way we can potentially reduce the amount of
VFIO PCI quirks that are built in.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: vfio: quirks: Split NVIDIA quirks out

And provide stubs when VFIO_QUIRKS_NVIDIA is not set.
This way we can potentially reduce the amount of
VFIO PCI quirks that are built in.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

yangzhon commented:
@sameo ,

I have merged the VFIO full build patch into your patch, so you don't need change any more.
I have removed WIP from this PR, please help review. thanks!

sameo commented on default-configs/hw-common.mak:
```diff
@@ -0,0 +1,3 @@
+CONFIG_BLOCK=y
```
`BLOCK` will be needed by everyone, I think we can remove that one.

yangzhon replied:
thanks, this file is built-in file for all platform. I will remove this CONFIG.

sameo commented on hw/misc/Makefile.objs:
```diff
@@ -9,7 +9,7 @@ common-obj-$(CONFIG_PCI_TESTDEV) += pci-testdev.o
 common-obj-$(CONFIG_PCA9552) += pca9552.o
 
-common-obj-y += unimp.o
+common-obj-$(CONFIG_UNIMP_DEV) += unimp.o
```
Could we please rename that one to `CONFIG_UNIMPLEMENTED` ?

yangzhon replied:
Done, thanks!

sameo commented on hw/nvram/Makefile.objs:
```diff
@@ -1,7 +1,7 @@
 common-obj-$(CONFIG_DS1225Y) += ds1225y.o
-common-obj-y += eeprom93xx.o
+common-obj-$(CONFIG_93XX_EEPROM) += eeprom93xx.o
```
Only 2 devices are depending on this one: esp-pci and eepro100. So we should use:

```
common-obj-$(call lor,$(CONFIG_EEPRO100_PCI),$(CONFIG_ESP_PCI)) += eepprom93xx.o
```

And  avoid carrying an additional `CONFIG_` knob.

yangzhon replied:
Done, thanks for review.

yangzhon replied:
@sameo 
when i used common-obj-$(call lor,$(CONFIG_EEPRO100_PCI),$(CONFIG_ESP_PCI)) += eepprom93xx.o

and do full build, the arm will report below issue
make: *** No rule to make target 'hw/nvram/eepprom93xx.o', needed by 'subdir-aarch64-softmmu'.  Stop.

In fact, CONFIG_EEPRO100_PCI and CONFIG_ESP_PCI have been defined in device.mak file

sameo commented on hw/nvram/Makefile.objs:
```diff
@@ -1,7 +1,7 @@
 common-obj-$(CONFIG_I2C) += eeprom_at24c.o
 common-obj-y += fw_cfg.o
-common-obj-y += chrp_nvram.o
+common-obj-$(CONFIG_NVRAM) += chrp_nvram.o
```
Same for that one, it's only used by a few devices/platforms. And avoid creating `CONFIG_NVRAM`.

yangzhon replied:
@sameo ,

Mac_nvram.c (hw\nvram):    sysp_end = chrp_nvram_create_system_partition(&nvr->data[off],
Spapr_nvram.c (hw\nvram):        chrp_nvram_create_system_partition(nvram->buf, MIN_NVRAM_SIZE / 4);
Sun4m.c (hw\sparc):    sysp_end = chrp_nvram_create_system_partition(image, 0);
Sun4u.c (hw\sparc64):    sysp_end = chrp_nvram_create_system_partition(image, 0);

Sun4m.c and Sun4u.c are all built-in files, if chrp_nvram.c base on platform, we need extra two CONFIGs for this.

sameo replied:
Ok, then let's keep that CONFIG but rename it to `CONFIG_CHRP_NVRAM`. And add it to the sparc and sparc64 .mak files.

yangzhon replied:
Done, thanks!

sameo commented on hw/acpi/Makefile.objs:
```diff
@@ -1,5 +1,7 @@
 ifeq ($(CONFIG_ACPI),y)
-common-obj-$(CONFIG_ACPI_X86) += core.o piix4.o pcihp.o
+common-obj-$(CONFIG_ACPI_X86) += core.o
+common-obj-$(CONFIG_ACPI_X86_PIXX) += piix4.o
```
Typo: `CONFIG_ACPI_X86_PIIX`, not `PIXX`

yangzhon replied:
It's my stupid error, thanks!

sameo commented on default-configs/mips-softmmu-common.mak:
```diff
@@ -36,3 +37,5 @@ CONFIG_EMPTY_SLOT=y
 CONFIG_MIPS_CPS=y
 CONFIG_MIPS_ITU=y
 CONFIG_I2C=y
+CONFIG_ACPI_X86_PIXX=y
```
Why does MIPS need an x86 specific ACPI configuration?

yangzhon replied:
@sameo ,

Please check below functions,
Mips_malta.c (hw\mips):    smbus = piix4_pm_init(pci_bus, piix4_devfn + 3, 0x1100,
Pc.h (include\hw\i386):I2CBus *piix4_pm_init(PCIBus *bus, int devfn, uint32_t smb_io_base,
Pc_piix.c (hw\i386):        smbus = piix4_pm_init(pci_bus, piix3_devfn + 3, 0xb100,
Piix4.c (hw\acpi):I2CBus *piix4_pm_init(PCIBus *bus, int devfn, uint32_t smb_io_base,


mips call piix4_pm_init(), which is defined in PIIX code, thanks!

yangzhon replied:
@sameo 

Sorry, i made a mistake. mips called  piix4_pm_init(), which is defined in Pc_piix.c, not Piix4.c.

So CONFIG_ACPI_X86_PIXX=y need remove, thanks!

sameo replied:
I see that the Malta boards are running with a piix4 chipset(!), my bad.
So then it would feel better to add `CONFIG_X86_PIIX` to the MIPS Malta config, and then `land($CONFIG_ACPI_X86, $CONFIG_X86_PIIX) in the `acpi/Makefile.objs` makefile for the piix4.o object.

yangzhon replied:
Done, thanks!

sameo commented on hw/cpu/Makefile.objs:
```diff
@@ -2,4 +2,4 @@ obj-$(CONFIG_ARM11MPCORE) += arm11mpcore.o
 obj-$(CONFIG_A9MPCORE) += a9mpcore.o
 obj-$(CONFIG_A15MPCORE) += a15mpcore.o
-common-obj-y += core.o
+common-obj-$(CONFIG_ARM_CORE) += core.o
```
Maybe this file is only used by ARM builds, but there's nothing ARM specific in there.
Having a more generic configuration name like `CONFIG_CPU_DEVICE` would be better.

yangzhon replied:
Done, thanks!

sameo reviewed (changes requested):
Carried some of Rob's comments from PR #57

sameo commented on default-configs/x86_64_virt-softmmu.mak:
```diff
@@ -13,7 +13,6 @@ CONFIG_VIRTIO_9P=y
 CONFIG_VIRTIO_BALLOON=y
 CONFIG_VIRTIO_BLK=y
 CONFIG_VIRTIO_CRYPTO=y
-CONFIG_VIRTIO_GPU=y
```
Could we please squash that one into e4e0fe01b9c994a4deeacf0d53fdf75edf302217 ?

yangzhon replied:
@sameo,

Done, thanks!

yangzhon commented:
@sameo @rbradford @mcastelino

name "apic", no-user
name "ioapic", bus System, no-user
name "kvm-apic", no-user
name "kvm-ioapic", bus System, no-user

The IRQ related device model, we will offload the IRQs emulation to KVM side, so at least "IOAPIC" is not needed anymore.   I will do this in next NEW PR.

yangzhon committed:
hw: make misc build configurable

Add new CONFIG_UNIMPLEMENTED for misc/unimp.c and make
it build configurable.
UNIMPLEMENTED is exclusively used by ARM boards, so we add it to the
default ARM softmmu configuration file.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make pcmcia build configurable

Add new CONFIG_PCMCIA for pcmcia and make it configurable.
Only ARM machine types use PCMCIA so we add CONFIG_PCMCIA to the ARM
softmmu default configuration file.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make Bluetooth build configurable

We add a new CONFIG_BT for the hw/bt directory to be configurable. When
CONFIG_BT is off, all QEMU Bluetooth code is no longer built. Since
Bluetooth is not a platform specific feature, and we don't want to add
yet another *.mak (bluetooth.mak) file, CONFIG_BT is a configure option,
defaulting to yes.

CONFIG_BLUEZ becomes CONFIG_BT_BLUEZ and will be set if CONFIG_BT is set
as well.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make HID/PS2 build configurable

By adding the CONFIG_HID and CONFIG_PS2 configuration knobs
for hid and ps2 devices, we can compile them out.

HID and PS2 are not features that are directly pulled in from platforms,
and since QEMU lacks a way to describe configuration dependencies, we
need to add CONFIG_PS2 to all platforms that need it for their keyboard
emulation. CONFIG_HID is a USB abd BT requirement so we add it to
usb.mak and as a BT requirement from the configure script, respectively.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make nvram build configurable

Add new CONFIG_CHRP_NVRAM for the Common Hardware Reference Platform
(CHRP) NVRAM helpers.
They're only used by softmmu PPC, PPC64, SPARC and SPARC64 machine types.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: acpi: make machine type files configurable

Add a couple of ACPI configuration knobs so that we can
remove the piix4 and the hotplug objects from the default build.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: i386: make machine type code configurable

With the new X86_COMMON, X86_PIIX and X86_Q35 configuration knobs, we
can start removing the legacy machine types from the default build.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: input : make the virtio input device configurable

Split out the virtio input device from the hw/virtio/virtio-pci.c
and make those virtio input devices configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: nvdimm: make nvdimm configurable

Make nvdimm configurable and add stub functions if we compile
out nvdimm.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: core: make register configurable

register.c is only used by xlinx and make this configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: core: make gpio irq configurable

or-irq and split-irq are only used for supporting GPIO IRQs,
there is no reason to include them by default.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

sameo commented on hw/vfio/stub.c:
```diff
@@ -0,0 +1,25 @@
+/*
+ * QEMU i386 virt Platform stub
```
This is the VFIO stubs, not the virt platform one.

yangzhon replied:
Done, thanks!

yangzhon committed:
audio: Make audio optional and configurable

CONFIG_AUDIO is now the top-level AUDIO configuration knob, and is
disabled through a new --disable-audio configure option.
By default it is enabled.

When disabled, all the hw/audio and audio/ code is replaced with a pair
of audio stub files.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make pci_bridge_dev.c build configurable

Add new CONFIG for pci_bridge_dev.c and make it
build configurable. By default it is enabled for all platforms
including pci.mak

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make wdt build configurable

Add CONFIG for watchdog.c in hw/watchdog and make
it build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make shpc.c build configurable

Add the new CONFIG_PCI_SHPC for pci/shpc.c and make
it build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon commented:
@sameo ,

I addressed all your comments and please review again, thanks!

sameo reviewed (approved):
LGTM

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -1,3 +1,4 @@
 CONFIG_CDROM=y
 CONFIG_HARD_DISK=y
 CONFIG_UNIMPLEMENTED=y
+CONFIG_PCMCIA=y
```
There is only one file inside pcmcia/ that is not trivial and that is controlled by CONFIG_PXA2XX which is only turned on in arm-softmmu.mk.

So please move CONFIG_PCMCIA=y to that file instead

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -2,3 +2,4 @@ CONFIG_CDROM=y
 CONFIG_HARD_DISK=y
 CONFIG_UNIMPLEMENTED=y
 CONFIG_PCMCIA=y
+CONFIG_BT=y
```
Have you considered piggy-backing on the back of CONFIG_BLUEZ instead?

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -3,3 +3,5 @@ CONFIG_HARD_DISK=y
 CONFIG_PCMCIA=y
 CONFIG_BT=y
+CONFIG_HID_DEV=y
+CONFIG_PS2_DEV=y
```
Have you looked at adding these only to those targets that actually use them?

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -5,3 +5,4 @@ CONFIG_PCMCIA=y
 CONFIG_BT=y
 CONFIG_HID_DEV=y
 CONFIG_PS2_DEV=y
+CONFIG_CHRP_NVRAM=y
```
This is not a virtual device - just some helper functions. It's only used by a couple of sun machines and a couple of nvram devices. I think you can be more specific than putting it in this file

rbradford replied:
Also the commit message is wrong.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -6,3 +6,4 @@ CONFIG_BT=y
 CONFIG_HID_DEV=y
 CONFIG_PS2_DEV=y
 CONFIG_CHRP_NVRAM=y
+CONFIG_AUDIO=y
```
This is not a virtual device.

Compiling this out will break vl.c.

rbradford commented on hw/block/Makefile.objs:
```diff
@@ -1,4 +1,6 @@
-common-obj-y += block.o cdrom.o hd-geometry.o
+common-obj-y += block.o
+common-obj-$(CONFIG_CDROM) += cdrom.o
```
This is only used by two devices. It would be better to do OR on CONFIG_IDE_CORE and CONFIG_SCSI.

This would also remove the need to do the stub stuff you do later.

rbradford commented on hw/block/Makefile.objs:
```diff
@@ -1,4 +1,6 @@
-common-obj-y += block.o cdrom.o hd-geometry.o
+common-obj-y += block.o
+common-obj-$(CONFIG_CDROM) += cdrom.o
+common-obj-$(CONFIG_HARD_DISK) += hd-geometry.o
```
I don't see the benefit of making this file conditional. You still needed to include it on the purely virt build...

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -7,3 +7,4 @@ CONFIG_HID_DEV=y
 CONFIG_PS2_DEV=y
 CONFIG_CHRP_NVRAM=y
 CONFIG_AUDIO=y
+CONFIG_PCI_BRIDGE=y
```
Maybe only turn it on in makefiles that have CONFIG_PCI?

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -8,3 +8,4 @@ CONFIG_PS2_DEV=y
 CONFIG_CHRP_NVRAM=y
 CONFIG_AUDIO=y
 CONFIG_PCI_BRIDGE=y
+CONFIG_WDT=y
```
I thought watchdog was in the configure script?

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -9,3 +9,5 @@ CONFIG_CHRP_NVRAM=y
 CONFIG_AUDIO=y
 CONFIG_PCI_BRIDGE=y
 CONFIG_WDT=y
+CONFIG_VFIO_VGA=y
```
I think this should be in pci.mak instead.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -9,3 +9,5 @@ CONFIG_CHRP_NVRAM=y
 CONFIG_PCI_BRIDGE=y
 CONFIG_WDT=y
+CONFIG_VFIO_VGA=y
+CONFIG_RAMFB=y
```
This is only used by ARM virt, PC and Q35 so this can be in more specific files.

rbradford commented on default-configs/x86_64_virt-softmmu.mak:
```diff
@@ -1,71 +1,41 @@
+CONFIG_VIRTIO_SERIAL=y
+CONFIG_VIRTIO_PCI=y
 
-include pci.mak
```
I'd rather see this commit add the new file in its minimal form. Rather than have the two previous commits that add it and then slightly reduce it.

rbradford commented on default-configs/x86_64_virt-softmmu.mak:
```diff
@@ -22,8 +22,6 @@ CONFIG_ACPI=y
 CONFIG_APIC=y
 CONFIG_IOAPIC=y
-CONFIG_NVDIMM=y
-CONFIG_ACPI_NVDIMM=y
```
I don't think this should be deleted from this file at the moment. NVDIMM adds a lot of value to the virt platform.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -11,3 +11,4 @@ CONFIG_WDT=y
 CONFIG_VFIO_VGA=y
 CONFIG_RAMFB=y
 CONFIG_NULL_MACHINE=y
+CONFIG_REGISTER=y
```
In your commit message you say this is only used on Xilinx. So why does it need to be included in every build?

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -12,3 +12,4 @@ CONFIG_VFIO_VGA=y
 CONFIG_RAMFB=y
 CONFIG_NULL_MACHINE=y
 CONFIG_REGISTER=y
+CONFIG_GPIO_IRQ=y
```
I think these files are only used on ARM.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -0,0 +1,15 @@
+CONFIG_NULL_MACHINE=y
+CONFIG_REGISTER=y
+CONFIG_GPIO_IRQ=y
+CONFIG_USB_CORE=y
```
As you can see from my comments. My gut feeling is by being a little bit more careful we can remove this file entirely.

Commit:
hw: bt: Move top-level Bluetooth files under hw/bt

And remove dead code while doing so.
Having top level Bluetooth related code is a little awkward, given the
unusual use cases that this implements. Moving them under hw/bt makes
the code layout cleaner.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: cpu: Make the CPU core abstract device configurable

This is only needed by a few PPC64 and ARM machine types, there is no
reason to include it in every build.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

sameo commented:
This is ready to be merged.
With this configure script:

```
$SRCDIR/configure \
 --disable-fdt \
 --disable-libiscsi \
 --disable-libnfs \
 --disable-libssh2 \
 --disable-linux-aio \
 --disable-lzo \
 --disable-modules \
 --disable-netmap \
 --disable-qom-cast-debug \
 --disable-seccomp \
 --disable-snappy \
 --disable-tcmalloc \
 --disable-tools \
 --disable-tpm \
 --disable-virtfs \
 --disable-tcg \
 --disable-slirp \
 --disable-capstone \
 --disable-xen \
 --disable-wdt \
 --disable-audio \
 --disable-bluetooth \
 --disable-usb-redir \
 --enable-attr \
 --enable-cap-ng \
 --enable-kvm \
 --enable-rbd \
 --enable-vhost-crypto \
 --enable-vhost-net \
 --enable-vhost-scsi \
 --enable-vhost-user \
 --enable-vhost-vsock \
 --enable-virtfs \
 --target-list=x86_64_virt-softmmu \
 --extra-cflags="$EXTRA_CFLAGS" \
 --extra-ldflags=" -pie -z noexecstack -z relro -z now" \
 --libdir=/usr/lib64/nemu \
 --libexecdir=/usr/libexec/nemu \
 --datadir=/usr/share/nemu
```

I can generate a list of unused files and automatically remove them to get down to the following code size:

```
-------------------------------------------------------------------------------
Language                 Files     Lines     Code  Comments   Blanks Complexity
-------------------------------------------------------------------------------
C                          539    321651   245840     33774    42037      37683
C Header                   524     91249    51389     29102    10758        977
```

down from 

```
-------------------------------------------------------------------------------
Language                 Files     Lines     Code  Comments   Blanks Complexity
-------------------------------------------------------------------------------
C                         1824   1313592  1035283    119525   158784     131849
C Header                  1429    245153   161133     55763    28257       3557
```

rbradford commented on default-configs/alpha-softmmu.mak:
```diff
@@ -2,6 +2,7 @@
 
 include pci.mak
 include usb.mak
+include hw-common.mak
```
Since this file is empty. Maybe we could remove it from all the other files?

sameo replied:
Ah sorry, I meant to remove those but forgot.

sameo replied:
Done.

`git grep hw-common.mak` is empty now.

rbradford commented on default-configs/x86_64_virt-softmmu.mak:
```diff
@@ -0,0 +1,37 @@
+CONFIG_ACPI_MEMORY_HOTPLUG=y
+CONFIG_ACPI_CPU_HOTPLUG=y
+CONFIG_ACPI_PCI_HOTPLUG=y
+
```
Can you revert the change that @yangzhon did to remove CONFIG_NVDIMM=y from here? I think we want to keep it in this file.

sameo replied:
Yes, makes sense.

sameo replied:
Done as well.

rbradford commented on default-configs/tricore-softmmu.mak:
```diff
@@ -0,0 +1 @@
+include hw-common.mak
```
This looks strange? Has this changed because of the rebase?

sameo replied:
That's an odd one. It's untouched now.

rbradford commented on configure:
```diff
@@ -1112,6 +1120,10 @@ for opt do
+  --disable-bluetooth) bluetooth="no"
+  ;;
+  --enable-bluetooth) bluetooth="yes"
+  ;;
```
How is this different to the bluez version? Is there a !bluez backend?

sameo replied:
Yes, the BlueZ code is optional, you can have BT on without BlueZ support from the host.

However the BlueZ code is not relevant when BT is off. So you can have CONFIG_BT && CONFIG_BLUEZ, CONFIG_BT && !CONFIG_BLUEZ but not !CONFIG_BT && CONFIG_BLUEZ.

Commit:
tools: Add a virt specific configure and build script

With all unneeded emulated devices configured out.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

rbradford committed:
tools/CI: Make NATS scripts take parameters for go test

This allows subtest filtering which will be needed for virt only testing.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
build: Update Jenkinsfile for virt only build

Add infrastructure building and testing only virt machine builds.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
nats: Add support for specifying path to binary

When compiling the virt only target the binary name is different. Allow
controlling of that withing NATS.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>
